### PR TITLE
Save collections

### PR DESCRIPTION
**Pull request by jonathonherbert**
Save collections when article fragments are updated. This includes Reetta's work on the server endpoint. (It doesn't care about changes to collections ... yet!)

The thunk responsible for collection updates also mixes in updatedBy information. This could also live in the Collections reducer - this would probably be better! - but I thought I'd ship this first.

We add a middleware to our store that watches for actions with the following signature:

```
{
  ...
  meta: {
    // The resource to persist the data to
    persistTo: 'collection',
    // The key to take from the action payload. Defaults to 'id'.
    key?: string,
    // Should we find collection parents before or after the reducer is called?
    // This is important when the relevant collection is affected by when the operation
    // occurs - finding the parent collection before a remove operation, for example,
    // or after an add operation.
    applyBeforeReducer?: boolean
  }

}
```

If it sees an action with 'persistTo' with the value 'collection', it will dispatch a thunk to persist the affected collections(s). Actions with this meta mixed in to the default actions in 'shared' are available in src/Actions/Collections.

### CODE REVIEW

File: client-v2/src/util/storeMiddleware.js [hunk 1]
```diff
@@ -27,3 +32,109 @@ export const updateStateFromUrlChange: Middleware<State, Action> = ({
...
+   * @todo At the moment this just cares about updates to article fragments,
+   *   but it should also listen for edits to collections.
+   */
+  const getCollectionIdsForActions = (actions: Action[]) => {
+    const articleFragmentIds: string[] = uniq(
+      actions.map(
+        // A sneaky 'any' here, as it's difficult to handle dynamic key
+        // values with static action types.
+        (act: any) =>
+          act.payload && act.meta.persistTo && act.meta.key
```
Comment (RichieAHB) [inline client-v2/src/util/storeMiddleware.js]: How come we're checking for `persistTo` here - should it not always be set?

File: client-v2/src/util/storeMiddleware.js [hunk 2]
```diff
@@ -27,3 +32,109 @@ export const updateStateFromUrlChange: Middleware<State, Action> = ({
...
+    );
+
+    // Persist the lot!
+    const sharedState = selectSharedState(store.getState());
+    collectionIds.forEach(id => {
+      const collection = selectors.selectById(sharedState, id);
+      // Flow has problems with us dispatching thunks here.
+      // This relates to a problem with the middleware definition -
+      // see https://github.com/flowtype/flow-typed/issues/574
+      // $FlowFixMe
```
Comment (RichieAHB) [inline client-v2/src/util/storeMiddleware.js]: Could we cast dispatch to `any` and then back to our `Dispatch` here just for a sanity check?

File: client-v2/src/actions/ArticleFragments.js [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+} from 'shared/actions/ArticleFragments';
+
+/**
+ * We add the persistence middleware meta to actions we'd like to trigger persist
+ * operations here - see the persistCollectionOnEdit middleware.
+ */
+function removeSupportingArticleFragmentWithPersist(...args: *) {
+  return {
+    ...removeSupportingArticleFragment(...args),
+    meta: {
```
[the remark this answers is not in the document]
Comment (jonathonherbert) [reply client-v2/src/actions/ArticleFragments.js]: Yes, it is, that's much more elegant ❤️

File: client-v2/src/shared/util/shared.js [hunk 4]
```diff
@@ -126,4 +135,55 @@ const normaliseCollectionWithNestedArticles = (
...
+function denormaliseArticleFragment(state: State, id: string) {
+  let articleFragment = articleFragmentSelector(selectSharedState(state), id);
+  if (!articleFragment) {
+    throw new Error(
+      `Could not denormalise article fragment - no article fragment found with id '${id}'`
+    );
+  }
+  if (articleFragment.meta && articleFragment.meta.supporting) {
+    articleFragment = set(
+      ['meta', 'supporting'],
```
[the remark this answers is not in the document]
Comment (susiecoleman) [reply client-v2/src/shared/util/shared.js]: Maybe just trying to find a consistent way of doing it every time? I think it's hard to move to `supporting` without removing the meta level from the model as anyone new coming on to the project is unlikely to know/look for a helper function

File: client-v2/src/util/storeMiddleware.js [hunk 5]
```diff
@@ -27,3 +32,109 @@ export const updateStateFromUrlChange: Middleware<State, Action> = ({
...
+    applyBeforeReducer?: boolean
+  }
+|};
+
+/**
+ * Return an array of actions - either a single action,
+ * or multiple actions if the action contains batched actions.
+ */
+const unwrapBatchedActions = (action: ActionWithBatchedActions): Action[] =>
+  action.type === 'BATCHING_REDUCER.BATCH' ? action.payload : [action];
```
[the remark this answers is not in the document]
Comment (jonathonherbert) [reply client-v2/src/util/storeMiddleware.js]: I can't see an export and it makes me sad. But hopefully we'll be removing redux-batched-actions soon.

File: client-v2/src/util/storeMiddleware.js (hunk 1 above)
Comment (jonathonherbert) [reply client-v2/src/util/storeMiddleware.js]: redux-batched-actions provides a meta key too, so we must still check for the property.

**Pull request merged**